src/main/groovy/wooga/gradle/paket/base/utils/internal/UnwrapUpm.groovy

```diff
@@ -0,0 +1,52 @@
+package wooga.gradle.paket.base.utils.internal
+
+import org.apache.commons.io.FileUtils
+import org.gradle.api.GradleException
+import org.gradle.api.Project
+
+class UnwrapUpm {
+    private File upmTarFile;
+    private File destinationDir;
+
+    // this is how Unity names the root directory inside the upm tar files
+    private static final String unpackDirectoryName = "package"
+
+    UnwrapUpm(File tarFile, File destinationDir)
+    {
+        this.upmTarFile = tarFile
+        this.destinationDir = destinationDir;
+    }
+
+    UnwrapUpm(String url, File destinationDir, Project project)
+    {
+        this(project.file(url), destinationDir)
+    }
+
+    UnwrapUpm(PaketUPMWrapperReference ref, File outputDirectory)
+    {
+        this(ref.upmPackageTar, new File(outputDirectory, ref.upmPackageName))
+    }
+
+    public void exec(Project project)
+    {
+        if (!upmTarFile.exists()) {
+            throw new GradleException("No upm package tarball found in ${upmTarFile.path}")
+        }
+
+        project.logger.info("unpacking ${this.upmTarFile.path} into ${destinationDir.path}")
+
+        project.tarTree(this.upmTarFile).files.each { fileFromTar ->
+            def path = fileFromTar.toPath()
+            def indexOfPackage = 0
+            while (indexOfPackage < path.getNameCount() && path.getName(indexOfPackage).toString() != unpackDirectoryName) {
+                indexOfPackage++;
+            }
+
+            def relativePath = path.subpath(indexOfPackage+1, path.getNameCount())
+
+            project.logger.debug("copyfile ${fileFromTar.path} into ${new File(destinationDir, relativePath.toString()).path}")
+
+            FileUtils.copyFile(fileFromTar, new File(destinationDir, relativePath.toString()))
+        }
```

Comment:
Was a gradle copy spec not possible here? It is at least for us using gradle way more readable.
To be fair I don't quite understand the reason for the while loop to find a common base path. I think a small commend what you try to solve. 
The `Path` API has for instance `f.toPath().relativize(otherPath)`.

Otherwise I would have assumed this would be quicker

```
project.copy {
    from(upmTarFile)
    into(destinationDir)
}
```

Reply:
The while loop isn't just looking for a common base path, it is looking for the `package` directory (upm convention) from inside the tar, in order to copy that into destination dir. But I'll try to see if i can use the cooler DSL :)

Reply:
Ah I see. The copySpec api has some tricks up its sleeve but if this is in the end helpfull or mor readable ... :)

Reply:
that DSL only works in .gradle files, i'm in imperative groovy land